crates/bevy_ecs/src/world/unsafe_world_cell.rs

```diff
@@ -70,75 +72,164 @@ use std::any::TypeId;
+    /// so the returned `&World` should not be handed out to safe code and care should be taken
+    /// when working with it.
+    ///
+    /// Deliberately private as the correct way to access data in a [`World`] that may have existing
+    /// mutable borrows of data inside it, is to use [`UnsafeWorldCell`].
+    ///
+    /// # Safety
+    /// - must not be used in a way that would conflict with any
+    ///   live exclusive borrows on world data
+    unsafe fn unsafe_world(self) -> &'w World {
+        // SAFETY:
+        // - caller ensures that the returned `&World` is not used in a way that would conflict
+        //   with any existing mutable borrows of world data
+        unsafe { &*self.0 }
     }
 
     /// Retrieves this world's [Entities] collection
     #[inline]
     pub fn entities(self) -> &'w Entities {
-        &self.0.entities
+        // SAFETY:
+        // - we only access world metadata
+        &unsafe { self.unsafe_world() }.entities
     }
 
     /// Retrieves this world's [Archetypes] collection
     #[inline]
     pub fn archetypes(self) -> &'w Archetypes {
-        &self.0.archetypes
+        // SAFETY:
+        // - we only access world metadata
+        &unsafe { self.unsafe_world() }.archetypes
     }
 
     /// Retrieves this world's [Components] collection
     #[inline]
     pub fn components(self) -> &'w Components {
-        &self.0.components
-    }
-
-    /// Retrieves this world's [Storages] collection
-    #[inline]
-    pub fn storages(self) -> &'w Storages {
-        &self.0.storages
+        // SAFETY:
+        // - we only access world metadata
+        &unsafe { self.unsafe_world() }.components
     }
 
     /// Retrieves this world's [Bundles] collection
     #[inline]
     pub fn bundles(self) -> &'w Bundles {
-        &self.0.bundles
+        // SAFETY:
+        // - we only access world metadata
+        &unsafe { self.unsafe_world() }.bundles
     }
 
     /// Reads the current change tick of this world.
     #[inline]
     pub fn read_change_tick(self) -> u32 {
-        self.0.read_change_tick()
+        // SAFETY:
+        // - we only access world metadata
+        unsafe { self.unsafe_world() }
+            .change_tick
+            .load(Ordering::Acquire)
     }
 
     #[inline]
     pub fn last_change_tick(self) -> u32 {
-        self.0.last_change_tick()
+        // SAFETY:
+        // - we only access world metadata
+        unsafe { self.unsafe_world() }.last_change_tick
     }
 
     #[inline]
     pub fn increment_change_tick(self) -> u32 {
-        self.0.increment_change_tick()
+        // SAFETY:
+        // - we only access world metadata
+        unsafe { self.unsafe_world() }
+            .change_tick
+            .fetch_add(1, Ordering::AcqRel)
+    }
+
+    /// Shorthand helper function for getting the [`ArchetypeComponentId`] for a resource.
+    #[inline]
+    pub fn get_resource_archetype_component_id(
```

Comment:
```suggestion
    pub(crate) fn get_resource_archetype_component_id(
```

The function on `World` is also `pub(crate)`.

Reply:
I don't see a problem with this being public, probably ought to just make the world methods public too

Reply:
I'd just rather us not pollute the public interface with helper functions like these. There's very little practical use outside of bevy_ecs itself, and provides multiple ways of doing the same thing, which can be rather confusing.

Reply:
`SystemParam` impls and `WorldCell` both use this method, the point of `UnsafeWorldCell` is:
- allow users to write their own interior mutability abstractions around world
- clean up internals to make it clearer when we have a real `&World` and when we dont

the fact that we are using this method in most (all?) of our interior mutability abstractions makes it seem pretty reasonable to have it publicly available on the struct used for making interior mutability abstractions

Reply:
I'm still going to push back on making this public. There are other ways of fetching this information from `&World`, and IMO this should not be promoted as a common user-facing API. Even with UnsafeWorldCell operations as rare as they are, I'm not a fan of providing redundant shorthands in user-facing APIs unless it's saving users a lot of complexity, which I don't think this shorthand is doing.

Reply:
Fine. I've made it private because i cant be bothered to keep arguing this and I would like this to actually get merged